### PR TITLE
Add a plugin to use Perspective for monitoring battles

### PR DESCRIPTION
**Pull request by mia-pi-git**
To reviewers: Please do not merge, approve instead and I will merge post approval once I have the API quota handled.
tl;dr of what this does: 
- on each message in a rated battle, calls out to google's Perspective API to classify messages (in a child process)
- child process collates the data into a score, sends it back to the parent process, which caches it by the roomid and writes into the DB that it got a hit.
- If too many scored messages are found, it flags the room and updates a page that displays data on all flagged rooms.
For now, that's all it does - alerts staff. Limited to US at the moment, but will likely be expanded to lowerstaff soon.

Note that this is hotpatchable, but has a (non-required) monkeypatch to go with it.

### CODE REVIEW

**Comment by Slayer95** [comment]
Insults, threats, profanity and attacks are just plain "abusive chat". Nowadays "toxic" has actually become a buzzword, which imo should be reserved to describe more subtly harmful relationships.

**Comment by mia-pi-git** [comment]
> Perspective’s main attribute is TOXICITY, defined as “a rude, disrespectful, or unreasonable comment that is likely to make you leave a discussion”. 
Toxicity-monitor was a Zarel suggestion originally, but given this is how Perspective defines toxicity on their attribute page, I think the name is fine. It's certainly more concise than the other ideas we came up like "battle-chat-monitor"

**Comment by Zarel** [comment]
I don't mind `rudeness-monitor`. I think Slayer95 makes a good point that it's clearer and less "political" of a word than toxicity.

**Comment by mia-pi-git** [comment]
> I don't mind `rudeness-monitor`. I think Slayer95 makes a good point that it's clearer and less "political" of a word than toxicity.

OK. For posterity, we discussed in PMs and landed on `abuse-monitor` (more of an in-between).

File: server/chat-plugins/abuse-monitor.ts [hunk 1]
```diff
@@ -0,0 +1,555 @@
...
+			result[k] = score.summaryScore.value;
+		}
+		return result;
+	} catch (e) {
+		Monitor.crashlog(e, 'A Perspective API request', {request: JSON.stringify(request)});
+		return null;
+	}
+}
+
+export const PM = new ProcessManager.QueryProcessManager<any, any>(module, async ({comment}) => {
```
Comment (AnnikaCodes) [inline server/chat-plugins/abuse-monitor.ts:100]: Why is this typed `<any, any>`?

File: server/chat-plugins/abuse-monitor.ts [hunk 2]
```diff
@@ -0,0 +1,555 @@
...
+	if (message.startsWith('SLOW\n')) {
+		Monitor.slow(message.slice(5));
+	}
+});
+
+if (!PM.isParentProcess) {
+	// This is a child process!
+	global.Config = Config;
+	global.Monitor = {
+		crashlog(error: Error, source = 'A netfilter child process', details: AnyObject | null = null) {
```
Comment (AnnikaCodes) [inline server/chat-plugins/abuse-monitor.ts:139]:
```suggestion
		crashlog(error: Error, source = 'An abuse monitor child process', details: AnyObject | null = null) {
```

Comment (AnnikaCodes) [reply server/chat-plugins/abuse-monitor.ts:139]: "netfilter" makes me think of filtering network requests, not neural networks. If you want to allude to neural networks, what about nnfilter or mlfilter or neuralfilter?

File: server/chat-plugins/abuse-monitor.ts [hunk 3]
```diff
@@ -0,0 +1,555 @@
...
+			if (!cache[roomid]) cache[roomid] = {users: {}, flags: new Set()};
+			if (!cache[roomid].users[user.id]) cache[roomid].users[user.id] = 0;
+			cache[roomid].users[user.id] += score;
+			for (const flag of flags) cache[roomid].flags.add(flag);
+			let hitThreshold = 0;
+			if (cache[roomid].users[user.id] >= settings.threshold) {
+				cache[roomid].notified = true;
+				notifyStaff();
+				hitThreshold = 1;
+				void (room)?.uploadReplay?.(user, this.connection, "forpunishment");
```
Comment (AnnikaCodes) [inline server/chat-plugins/abuse-monitor.ts:176]:
```suggestion
				void room?.uploadReplay?.(user, this.connection, "forpunishment");
```

File: server/chat-plugins/abuse-monitor.ts [hunk 4]
```diff
@@ -0,0 +1,555 @@
...
+			}
+			if (settings.threshold === num) {
+				return this.errorReply(`The abuse monitor threshold is already ${num}.`);
+			}
+			settings.threshold = num;
+			saveSettings();
+			this.privateGlobalModAction(`${this.user.name} set the abuse monitor trigger threshold to ${num}.`);
+			this.globalModlog('ABUSEMONITOR THRESHOLD', null, `${num}`);
+			this.sendReply(
+				`|html|Remember to use <code>/tm respawn</code> to deploy the settings to the child process.`
```
Comment (AnnikaCodes) [inline server/chat-plugins/abuse-monitor.ts:258]:
```suggestion
				`|html|Remember to use <code>/am respawn</code> to deploy the settings to the child process.`
```

File: server/chat-plugins/abuse-monitor.ts [hunk 5]
```diff
@@ -0,0 +1,555 @@
...
+			const percent = parseFloat(rawPercent);
+			if (isNaN(percent) || percent > 1 || percent < 0) {
+				return this.errorReply(`Invalid percent: ${percent}. Must be between 0 and 1.`);
+			}
+			const score = parseInt(rawScore);
+			if ((isNaN(score) && rawScore !== 'MAXIMUM') || score < 0) {
+				return this.errorReply(`Invalid score: ${score}. Must be a positive integer or "MAXIMUM".`);
+			}
+			if (settings.specials[type]?.[percent] && !this.cmd.includes('f')) {
+				return this.errorReply(`That special case already exists. Use /tm forceeditspecial to change it.`);
```
Comment (AnnikaCodes) [inline server/chat-plugins/abuse-monitor.ts:352]: All the `/tm` references need to be updated.

File: server/chat-plugins/abuse-monitor.ts (hunk 2 above)
Comment (mia-pi-git) [reply server/chat-plugins/abuse-monitor.ts:139]: Oh, yeah. I just forgot to rename this.

File: server/chat-plugins/abuse-monitor.ts [hunk 6]
```diff
@@ -0,0 +1,555 @@
...
+
+import {FS, Utils, Net, ProcessManager, Repl} from '../../lib';
+import {Config} from '../config-loader';
+import {toID} from '../../sim/dex-data';
+
+const WHITELIST = ["mia"];
+// 20m. this is mostly here so we can use Monitor.slow()
+const PM_TIMEOUT = 20 * 60 * 1000;
+const ATTRIBUTES = {
+	'SEVERE_TOXICITY': {},
```
[the remark this answers is not in the document]
Comment (mia-pi-git) [reply server/chat-plugins/abuse-monitor.ts:19]: Dunno, mostly consistency.

[the remark this answers is not in the document]
Comment (mia-pi-git) [reply server/chat-plugins/abuse-monitor.ts]: Done.

File: server/chat-plugins/abuse-monitor.ts (hunk 5 above)
Comment (mia-pi-git) [reply server/chat-plugins/abuse-monitor.ts]: Done.

File: server/chat-plugins/abuse-monitor.ts (hunk 1 above)
Comment (mia-pi-git) [reply server/chat-plugins/abuse-monitor.ts]: Fixed.

**Comment by mia-pi-git** [comment]
Convenient timing - Perspective just granted my quota increase request. Ready to go whenever code review is done.

**Comment by AnnikaCodes** [comment]
What's with the binary blob `dump.rdb`?

**Comment by mia-pi-git** [comment]
> What's with the binary blob `dump.rdb`?

Ugh. Forgot it wasn't gitignored.

File: server/chat-plugins/abuse-monitor.ts [hunk 7]
```diff
@@ -0,0 +1,560 @@
...
+			const unspawned = await PM.respawn();
+			this.sendReply(`DONE. ${Chat.count(unspawned, 'processes', 'process')} unspawned.`);
+			this.addGlobalModAction(`${user.name} used /abusemonitor respawn`);
+		},
+		async userclear(target, room, user) {
+			checkAccess(this);
+			const {targetUsername, rest} = this.splitUser(target);
+			const targetId = toID(targetUsername);
+			if (!targetId) return this.parse(`/help abusemonitor`);
+			if (user.lastCommand !== `tm userclear ${targetId}`) {
```
Comment (AnnikaCodes) [inline server/chat-plugins/abuse-monitor.ts:295]: more `tm` references

Comment (mia-pi-git) [reply server/chat-plugins/abuse-monitor.ts:295]: Blegh.

Comment (AnnikaCodes) [reply server/chat-plugins/abuse-monitor.ts:295]: Though I guess it's technically OK since it's not user-facing.

Comment (mia-pi-git) [reply server/chat-plugins/abuse-monitor.ts:295]: Ok, fixed.

**Review by AnnikaCodes — APPROVED** [review]
This seems OK to me.

**Comment by Slayer95** [comment]
@monsanto you may want to ``git filter-repo``
